src/history.rs

```diff
@@ -155,126 +153,184 @@ impl<'a> DynamicHistory<'a> {
-        let mut compact_log = CompactLogVec::from_bytes(log_region);
-        if clocks.capacity() < MIN_CLOCKS_LEN || compact_log.capacity() < MIN_LOG_LEN {
+        // Create new racebuffer, using full log region instead of rounding to power of 2 length for
+        // optimized indexing
+        // Note: point of future improvement - a heuristic could be used to determine whether or not the memory cost
+        // of rounding the log's storage space outweighs the runtime cost of using mod operations for indexing
+        let log = RaceLog::new_from_bytes(log_region, false)
+            .map_err(|_| StorageSetupError::UnderMinimumAllowedSize)?;
+        if clocks.capacity() < MIN_CLOCKS_LEN || (log.capacity() as usize) < MIN_LOG_LEN {
             return Err(StorageSetupError::UnderMinimumAllowedSize);
         }
         clocks
             .try_push(LogicalClock {
                 id: probe_id,
-                epoch: 0,
-                ticks: 0,
+                epoch: ProbeEpoch(0),
+                ticks: ProbeTicks(0),
             })
             .expect(
                 "The History.clocks field should always contain a clock for this probe instance",
             );
-        // This ensures that the first log segment always has a piece of logical
-        // clock information.
-        DynamicHistory::write_clocks_to_log(&mut compact_log, clocks.as_slice());
-        Ok(DynamicHistory {
+        let history = DynamicHistory {
+            read_cursor: 0,
+            report_seq_num: 0,
+            event_count: 0,
+            self_clock: LogicalClock {
+                id: probe_id,
+                epoch: ProbeEpoch(0),
+                ticks: ProbeTicks(0),
+            },
             probe_id,
             clocks,
-            compact_log,
-            chunked_report_state: ChunkedReportState::default(),
-            event_count: 0,
-        })
+            log,
+        };
+        Ok(history)
+    }
+
+    #[inline]
+    fn merge_overwritten_clock(&mut self, overwritten: OverwrittenEntry<LogEntry>) {
+        if let OverwrittenEntry::Double(one, two) = overwritten {
+            // If what we get out of the log is a clock, merge it into clocks list
+            if one.has_clock_bit_set() {
+                if let Some(id) = ProbeId::new(one.interpret_as_logical_clock_probe_id()) {
+                    let (epoch, ticks) = crate::unpack_clock_word(two.raw());
+                    self.merge_clock(LogicalClock { id, epoch, ticks });
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn write_to_log(&mut self, entry: LogEntry) {
+        let overwritten = self.log.write(entry);
+        self.merge_overwritten_clock(overwritten);
+    }
+
+    #[inline]
+    fn write_double_to_log(&mut self, prefix: LogEntry, suffix: LogEntry) {
+        let (first_overwritten, second_overwritten) = self.log.write_double(prefix, suffix);
+        self.merge_overwritten_clock(first_overwritten);
```

Comment:
These entries might not be clocks, and we should know this at the call site.  I would rather we have `write_event_with_payload_to_log` and `write_clock_pair_to_log` than tossing both pair-types to a method that's named for just clocks (and conducts a redundant check inside).